metadata/src/main/java/org/apache/kafka/controller/LogReplayTracker.java

```diff
@@ -17,56 +17,516 @@
+
+        try {
+            gainLeadershipCallback.accept(newRaftLeader.epoch());
+        } catch (Throwable e) {
+            fatalFaultHandler.handleFault("exception while claiming leadership", e);
+        }
+    }
+
+    private void becomeInactive(int epoch) {
+        this.active = false;
+        this.nextWriteOffset = 0L;
+        this.curClaimEpoch = -1;
+        controllerMetrics.setActive(false);
+        revert(lastCommittedOffset, "becoming inactive");
+        try {
+            loseLeadershipCallback.accept(epoch);
+        } catch (Throwable e) {
+            fatalFaultHandler.handleFault("exception in loseLeadership callback", e);
+        }
+    }
+
+    /**
+     * Find the latest offset that is safe to perform read operations at.
+     *
+     * @return The offset.
+     */
+    long lastStableOffset() {
+        // Calculate what the stable offset would be if metadata transactions did not exist.
+        //
+        // On the active controller, this is the lastCommittedOffset. On the standby controller,
+        // it is a special value indicating that we should always read the very latest data from
+        // all data structures. The reason for this difference is that the standby never has
+        // uncommitted data in memory. Using lastCommittedOffset for the standby would not work
+        // because the standby (unlike the active) doesn't keep in-memory snapshots at every
+        // committed offset at all times.
+        long stableOffsetIgnoringTransactions = active ?
+                lastCommittedOffset :
+                SnapshotRegistry.LATEST_EPOCH;
+
+        if (transactionStartOffset == -1) {
+            // If we're not inside a transaction, use the offset calculated above.
+            return stableOffsetIgnoringTransactions;
+        } else {
+            // If we're inside a transaction and it started before the above offset, use the
+            // transaction start offset instead.
+            return Math.min(stableOffsetIgnoringTransactions, transactionStartOffset);
+        }
+    }
+
+    LeaderAndEpoch raftLeader() {
+        return raftLeader;
+    }
+
+    long nextWriteOffset() {
+        if (!active) {
+            throw new RuntimeException("Cannot access the next write offset when we are not active.");
+        }
+        return nextWriteOffset;
+    }
+
+    QuorumControllerMetrics controllerMetrics() {
+        return controllerMetrics;
+    }
+
+    void replay(BeginTransactionRecord record, long offset) {
+        if (transactionStartOffset != -1L) {
+            throw fatalFaultHandler.handleFault("tried to begin metadata transaction at " + offset +
+                ", but we are still inside a metadata transaction starting at " +
+                transactionStartOffset);
+        }
+        log.debug("Beginning metadata transaction at offset {}{}.", offset,
+            record.name() == null ? "" : " named " + record.name());
+        snapshotRegistry.getOrCreateSnapshot(offset);
```

Comment:
Is it possible to move all the SnapshotRegistry mutating calls (getOrCreateSnapshot, deleteSnapshots, revertToSnapshot) into LogReplayTracker? It's a little hard to follow the in-memory snapshotting when it's split between here and QuorumController?